File: ios/GrMob/App/SystemEvents.swift
import UIKit

/// The iOS half of grmob's system-event channel: the app→host events that are
/// deliberately not part of the view tree, because what they reach is the
/// platform itself rather than anything the reconciler could diff.
///
///     core.ShowToast  ──▶ "toast"     ──▶ a transient overlay label
///     core.OpenURL    ──▶ "open_url"  ──▶ UIApplication.open
///
/// Before this existed the events were emitted into a nil Go handler and
/// vanished on both natives — only the WASM host had a sink — so an app
/// calling ShowToast worked in the browser preview and silently did nothing
/// on a device.
///
/// Unknown event names are dropped, matching the contract every other host
/// applies: a newer app on an older shell degrades to silence rather than to
/// a crash.
enum SystemEvents {
    /// Wires Go's event stream to this process's UI.
    ///
    /// Call before `GrMobRuntime.start()` so an event emitted during the very
    /// first render pass has somewhere to land.
    static func attach(_ bridge: GrMobBridge) {
        bridge.setSystemEventListener { name, payload in
            // The callback runs on the Go goroutine that emitted the event.
            // Everything below is UIKit, which is main-actor only, so every
            // event hops — the same hop GrMobRuntime makes for patches. Main
            // queue dispatches run FIFO, so a toast emitted before a
            // navigation still appears first.
            Task { @MainActor in
                dispatch(name: name, payload: payload)
            }
        }
    }

    @MainActor
    private static func dispatch(name: String, payload: String) {
        guard let data = payload.data(using: .utf8),
              let object = try? JSONSerialization.jsonObject(with: data) as? [String: Any]
        else {
            NSLog("GrMob: malformed payload for system event '\(name)'")
            return
        }
        switch name {
        case "toast": showToast(object)
        case "open_url": openURL(object)
        default: break
        }
    }

    @MainActor
    private static func openURL(_ data: [String: Any]) {
        guard let string = data["url"] as? String, !string.isEmpty,
              let url = URL(string: string)
        else { return }
        // `canOpenURL` is deliberately not consulted first: since iOS 9 it
        // answers only for schemes declared in LSApplicationQueriesSchemes,
        // so it reports false for perfectly openable custom schemes and would
        // suppress links that work. `open` reports the truth in its
        // completion handler, and there is nothing to do with the answer —
        // core.OpenURL is fire-and-forget by contract and a system event has
        // no return channel — so it is logged and dropped.
        UIApplication.shared.open(url, options: [:]) { opened in
            if !opened { NSLog("GrMob: nothing could open \(string)") }
        }
    }

    // MARK: - Toast

    /// The toast currently on screen, if any. A second toast replaces the
    /// first rather than stacking: the payload carries no identity, so there
    /// is no way to lay two of them out relative to each other, and a queue
    /// would make a burst of them outlast the action that caused it.
    @MainActor private static var current: UIView?

    @MainActor
    private static func showToast(_ data: [String: Any]) {
        guard let message = data["message"] as? String, !message.isEmpty else { return }
        // Go sends milliseconds (core.ToastConfig defaults to 2000). Unlike
        // Android's two fixed buckets, an overlay we draw ourselves can honor
        // the number exactly, so it does.
        let seconds = Double(data["duration"] as? Int ?? 2000) / 1000.0

        guard let window = keyWindow else { return }
        current?.removeFromSuperview()

        let label = PaddedLabel()
        label.text = message
        label.numberOfLines = 0
        label.textAlignment = .center
        label.textColor = .white
        label.font = .preferredFont(forTextStyle: .subheadline)
        // The payload's "style" key (core.UseToastStyle) is ignored: honoring
        // a core.Style here would mean a second, partial implementation of
        // GrMobStyle living outside the renderer. A styled toast degrades to
        // the platform look rather than to nothing.
        label.backgroundColor = UIColor.black.withAlphaComponent(0.82)
        label.layer.cornerRadius = 10
        label.layer.masksToBounds = true
        label.alpha = 0
        label.translatesAutoresizingMaskIntoConstraints = false
        // Transparent to touch: a toast is a notification, not a control, and
        // one covering the bottom of the screen must not eat taps meant for
        // whatever is under it.
        label.isUserInteractionEnabled = false

        window.addSubview(label)
        NSLayoutConstraint.activate([
            label.centerXAnchor.constraint(equalTo: window.centerXAnchor),
            label.bottomAnchor.constraint(
                equalTo: window.safeAreaLayoutGuide.bottomAnchor, constant: -32),
            label.leadingAnchor.constraint(
                greaterThanOrEqualTo: window.leadingAnchor, constant: 24),
            label.trailingAnchor.constraint(
                lessThanOrEqualTo: window.trailingAnchor, constant: -24),
        ])
        current = label

        UIView.animate(withDuration: 0.2) { label.alpha = 1 }
        UIView.animate(withDuration: 0.3, delay: seconds, options: []) {
            label.alpha = 0
        } completion: { _ in
            label.removeFromSuperview()
            if current === label { current = nil }
        }
    }

    @MainActor
    private static var keyWindow: UIWindow? {
        UIApplication.shared.connectedScenes
            .compactMap { $0 as? UIWindowScene }
            .flatMap(\.windows)
            .first { $0.isKeyWindow }
    }
}

/// A UILabel with interior padding. UILabel has no content inset, and a toast
/// whose text touches its rounded corners reads as a bug.
private final class PaddedLabel: UILabel {
    private let insets = UIEdgeInsets(top: 10, left: 16, bottom: 10, right: 16)

    override func drawText(in rect: CGRect) {
        super.drawText(in: rect.inset(by: insets))
    }

    override var intrinsicContentSize: CGSize {
        let size = super.intrinsicContentSize
        return CGSize(
            width: size.width + insets.left + insets.right,
            height: size.height + insets.top + insets.bottom)
    }

    // Multi-line labels lay out against a preferredMaxLayoutWidth derived
    // from the bounds, so the insets have to come off that too or the last
    // line runs under the right padding.
    override func textRect(
        forBounds bounds: CGRect, limitedToNumberOfLines numberOfLines: Int
    ) -> CGRect {
        let rect = super.textRect(
            forBounds: bounds.inset(by: insets), limitedToNumberOfLines: numberOfLines)
        return rect.inset(by: UIEdgeInsets(
            top: -insets.top, left: -insets.left,
            bottom: -insets.bottom, right: -insets.right))
    }
}

File: android/app/src/main/java/com/grmob/app/SystemEvents.kt
package com.grmob.app

import android.content.ActivityNotFoundException
import android.content.Context
import android.content.Intent
import android.net.Uri
import android.os.Handler
import android.os.Looper
import android.util.Log
import android.widget.Toast
import com.grmob.runtime.GrMobBridge
import org.json.JSONObject

/**
 * The Android half of grmob's system-event channel: the app→host events that
 * are deliberately not part of the view tree, because what they reach is the
 * platform itself rather than anything the reconciler could diff.
 *
 *   core.ShowToast  ──▶ "toast"     ──▶ android.widget.Toast
 *   core.OpenURL    ──▶ "open_url"  ──▶ Intent(ACTION_VIEW)
 *
 * Before this existed the events were emitted into a nil Go handler and
 * vanished on both natives — only the WASM host had a sink — so an app
 * calling ShowToast worked in the browser preview and silently did nothing on
 * a phone.
 *
 * Unknown event names are dropped, matching the contract every other host
 * applies: a newer app on an older shell degrades to silence rather than to a
 * crash.
 */
object SystemEvents {
    private const val TAG = "GrMobSystemEvents"

    /**
     * Wires Go's event stream to this process's UI.
     *
     * Call before [com.grmob.runtime.GrMobRuntime.start] so an event emitted
     * during the very first render pass has somewhere to land.
     *
     * @param context any Context; only its application context is retained,
     *   so an Activity passed here is not leaked past its own lifetime. The
     *   consequence is that [Intent.FLAG_ACTIVITY_NEW_TASK] is mandatory
     *   below — an application context has no task of its own to launch into.
     */
    fun attach(context: Context, bridge: GrMobBridge) {
        val appContext = context.applicationContext
        val main = Handler(Looper.getMainLooper())
        // The callback runs on the Go goroutine that emitted the event. Both
        // actions below touch the UI (a Toast must be shown from a Looper
        // thread; startActivity from an arbitrary thread is unreliable), so
        // every event is posted to the main thread — the same hop
        // GrMobRuntime makes for patches, and posts run FIFO, so a toast
        // emitted before a navigation still appears first.
        bridge.setSystemEventListener { name, payload ->
            main.post { dispatch(appContext, name, payload) }
        }
    }

    private fun dispatch(context: Context, name: String, payload: String) {
        val data = try {
            JSONObject(payload)
        } catch (e: org.json.JSONException) {
            Log.w(TAG, "malformed payload for system event '$name'", e)
            return
        }
        when (name) {
            "toast" -> showToast(context, data)
            "open_url" -> openUrl(context, data)
        }
    }

    private fun showToast(context: Context, data: JSONObject) {
        val message = data.optString("message")
        if (message.isEmpty()) return
        // Go sends milliseconds; Android offers two fixed buckets and honors
        // nothing in between (a custom duration has been ignored since API
        // 30). 3000ms is the midpoint between SHORT's ~2s and LONG's ~3.5s,
        // so the default 2000 lands on SHORT and an app that deliberately
        // asked for longer gets LONG.
        val duration = if (data.optInt("duration", 2000) > 3000) {
            Toast.LENGTH_LONG
        } else {
            Toast.LENGTH_SHORT
        }
        // A styled toast's "style" key is ignored: since API 30 the platform
        // owns the toast's appearance and setView is a no-op. Honoring it
        // would mean drawing our own overlay, which is a view-tree feature
        // and belongs in the renderer, not here.
        Toast.makeText(context, message, duration).show()
    }

    private fun openUrl(context: Context, data: JSONObject) {
        val url = data.optString("url")
        if (url.isEmpty()) return
        val intent = Intent(Intent.ACTION_VIEW, Uri.parse(url))
            .addFlags(Intent.FLAG_ACTIVITY_NEW_TASK)
        try {
            context.startActivity(intent)
        } catch (e: ActivityNotFoundException) {
            // No app claims the scheme — a `tel:` link on a tablet with no
            // dialer, or a device with no browser. Logged and swallowed:
            // core.OpenURL is fire-and-forget by contract and a system event
            // has no return channel, so there is nothing to report back to Go.
            Log.w(TAG, "no activity can open $url", e)
        }
    }
}
